nsot/models.py

```diff
@@ -1470,6 +1470,105 @@ def to_dict(self):
         }
 
 
+class Circuit(Resource):
+    """Represents two network Interfaces that are connected"""
+
+    # a_endpoint interface
+    a_endpoint = models.ForeignKey(
+        Interface, db_index=True, related_name='circuits_a',
+        null=False, verbose_name='A-side Interface',
+        help_text='Unique ID of interface at the A-side'
+    )
+
+    # z_endpoint interface
+    z_endpoint = models.ForeignKey(
+        Interface, db_index=True, related_name='circuits_z',
+        null=True, verbose_name='Z-side Interface',
+        help_text='Unique ID of interface at the Z-side'
+    )
+
+    # We are currently inferring the site_id from the parent (A-side) Interface
+    # in the .save() method
+    site = models.ForeignKey(
+        Site, db_index=True, related_name='circuits',
+        on_delete=models.PROTECT,
+        help_text='Unique ID of the Site this Circuit is under'
+    )
+
+    name = models.CharField(
+        max_length=255, unique=True,
+        help_text='Unique display name of circuit',
+        default='',
+    )
+
+    def __unicode__(self):
+        return u'%s' % self.name
+
+    class Meta:
+        # An interface shouldn't be able to be part of multiple circuits. This
+        # fixes the problem... mostly
+        #
+        # This doesn't solve an interface from being in one a_endpoint and
+        # another z_endpoint, but solves the others
+        unique_together = (
+            ('a_endpoint', 'z_endpoint'),
+            ('a_endpoint',),
+            ('z_endpoint',),
+        )
+        index_together = unique_together
+
+    @property
+    def interfaces(self):
+        intf_list = [self.a_endpoint, self.z_endpoint]
+        intf_list = [i for i in intf_list if i]
+        return intf_list
+
+    def clean_site(self, value):
+        """Always enforce that site is set."""
+        if value is None:
+            return self.a_endpoint.site_id
+
+        return value
+
+    def set_name(self):
+        if self.name:
+            return False
+
+        # Add display name of hostname:intf::hostname:intf
+        a_name = '{}:{}'.format(
+            self.a_endpoint.device.hostname,
+            self.a_endpoint.name,
+        )
+        if getattr(self, 'z_endpoint'):
+            z_name = '{}:{}'.format(
+                self.z_endpoint.device.hostname,
+                self.z_endpoint.name,
+            )
+        else:
+            z_name = 'None'
+
+        # '--' as delimeter because an intf name may contain '-'
+        self.name = '{}--{}'.format(a_name, z_name)
+        return True
+
+    def clean_fields(self, exclude=None):
+        self.site_id = self.clean_site(self.site_id)
+
+    def save(self, *args, **kwargs):
+        self.full_clean()
+        self.clean_fields()
```

Comment:
Calling `.clean_fields()` is redundant. It's called by `.full_clean()`by default.